fake_tensor.py

```diff
@@ -0,0 +1,171 @@
+import torch
+
+from base_tensor import BaseTensor
+from torch.testing._internal.common_utils import (
+    run_tests,
+    TestCase,
+)
+from torch.utils._pytree import tree_map
+from functools import partial
+from torch.fx.operator_schemas import normalize_function
+from typing import Union
+import unittest
+
+aten = torch.ops.aten
+
+
+# Meta tensors give you the ability to run PyTorch code without having to
+# actually do computation through tensors allocated on a `meta` device.
+# Because the device is `meta`, meta tensors do not model device propagation.
+# FakeTensor extends MetaTensors to also carry an additional `fake_device`
+# which tracks devices that would have been used.
+
+
+class FakeTensor(BaseTensor):
+    fake_device: torch.device
+
+    @staticmethod
+    def __new__(cls, elem, device):
+        return super().__new__(cls, elem)
+
+    def __init__(self, elem, device: Union[torch.device, str]):
+        # elem does not need to be recorded, because FakeTensor *is a* elem
+        assert elem.device.type == "meta"
+        device if isinstance(device, torch.device) else torch.device(device)
+        assert device.type != "meta"
+        self.fake_device = device
+
+    @staticmethod
+    def from_tensor(t):
+        existing_device = t.device
+        return FakeTensor(t.to(device="meta"), existing_device)
+
+    @property
+    def device(self):
+        return self.fake_device
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        # Run the original computation
+        r = super().__torch_dispatch__(func, types, args, kwargs)
+        kwargs = kwargs if kwargs else {}
+
+        def wrap(e, device):
+            # inplace ops can return fake tensors
+            if isinstance(e, torch.Tensor) and not isinstance(e, cls):
+                return FakeTensor(e, device)
+            else:
+                return e
+
+        # Pytorch device is kwarg-only, except for _pin_memory/pin_memory
+        # which is not yet supported by meta-tensors
+        assert (
+            func != aten._pin_memory.default and func != aten.pin_memory.default
+        ), f"NYI: {func}"
+
+        # if device is specified, use that
+        # not sure this is actually needed.. device only shows up in constructors
+        if kwargs.get("device", None):
+            return tree_map(partial(wrap, device=kwargs["device"]), r)
+
+        # operators which copy size from another tensor do not
+        # also take device from the size tensor
+        # other size_as operators are not builtin operators
+        if func == torch.ops.aten.resize_as_.default:
+            # TODO: https://github.com/pytorch/pytorch/pull/77182
+            _, new_kwargs = normalize_function(
+                func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
+            )
```

Comment:
Should we be recommending people generally use the FX normalize function? It still seems a bit different from how I personally feel idiomatic handling for functions should be done (via defining a function, and then hooking it up via some register decorator).

Reply:
I originally thought I was going to be handling more than just `resize_as_` (and I still might, e.g. when [this](https://github.com/pytorch/pytorch/blob/master/torch/autograd/_functions/tensor.py#L25) is removed). When you only care about one argument `input` it would be easier to handle these functions in bulk than separately defining each operator, and also means that if the signature changes you dont also need to change the signature of your function. 

Would also save having to specify defaults as in https://github.com/albanD/subclass_zoo/blob/main/python_meta_tensor.py#L99

And for functions intercepted at the `__torch_function__` level, could potentially handle some of the implicit conversions that currently before the kernel is invoked - e.g. things like broadcasting lists, and allowing `tensor(0)` for `int` arguments